### PR TITLE
Add remaining opcodes to arithmetics validation

### PR DESCRIPTION
**Pull request by atgoo**
Add validation rules for:
- OpIAddCarry
- OpISubBorrow
- OpUMulExtended
- OpSMulExtended

Includes some refactoring of old code.

### CODE REVIEW

File: source/val/validation_state.h [hunk 1]
```diff
@@ -358,6 +358,11 @@ class ValidationState_t {
       uint32_t id, uint32_t* num_rows, uint32_t* num_cols,
       uint32_t* column_type, uint32_t* component_type) const;
 
+  // Collects struct member types into |member_types|.
+  // Returns false if parsing fails (wrong syntaxis or not struct type).
```
Comment (dneto0) [inline source/val/validation_state.h]: "syntaxis"?

Comment (dneto0) [reply source/val/validation_state.h]: Also returns false if the struct has no members.

Also, it's probably ok that the vector is not updated in the failure case.  Probably better to say that in the method comment as well.

File: source/validate_arithmetics.cpp [hunk 2]
```diff
@@ -405,7 +412,51 @@ spv_result_t ArithmeticsPass(ValidationState_t& _,
...
+            << spvOpcodeString(opcode);
+
+      const uint32_t left_type_id = GetOperandTypeId(_, inst, 2);
+      const uint32_t right_type_id = GetOperandTypeId(_, inst, 3);
+
+      if (left_type_id != result_types[0] || right_type_id != result_types[0])
+        return _.diag(SPV_ERROR_INVALID_DATA)
+            << "Expected both operands to be of Result Type member type: "
+            << spvOpcodeString(opcode);
+
```
Comment (dneto0) [inline source/validate_arithmetics.cpp]: Please add TODO for IAddCarry ISubBorrow and UMulExtended: The result types members must have signedness of 0.

**Comment by dneto0** [comment]
Also please indicate if you have run the Vulkan CTS shaders through this extra functionality.

File: source/validate_arithmetics.cpp (hunk 2 above)
Comment (atgoo) [reply source/validate_arithmetics.cpp]: That is handled above
`  if (!_.IsUnsignedIntScalarType(result_types[0]) &&
       !_.IsUnsignedIntVectorType(result_types[0]))
        return _.diag(SPV_ERROR_INVALID_DATA)
            << "Expected Result Type struct member types to be unsigned "
            << "integer scalar or vector: " << spvOpcodeString(opcode);`

Comment (dneto0) [reply source/validate_arithmetics.cpp]: Ah.  I missed that reading the code too quickly.  I didn't quite see the if test against the opcode.

**Comment by dneto0** [comment]
I'll wait for the bots to go green.

**Comment by dneto0** [comment]
Whoops.  Still need the typo fix and updated method comment.

**Comment by dneto0** [comment]
Please fix the typo and method comment.

**Comment by atgoo** [comment]
Ready for review.

**Comment by dneto0** [comment]
Code looks ok.

But:

> Also please indicate if you have run the Vulkan CTS shaders through this extra functionality.

**Comment by dneto0** [comment]
I've verified this works with latest Vulkan CTS tests.

**Comment by dneto0** [comment]
Rebased and pushed into master as dcf42433a63c

**Pull request closed without merging**